### PR TITLE
🏃 [e2e] add clusterctl e2e

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR extends the e2e framework by adding an e2e friendly wrapper on top of the clusterctl client library; most notably, this can work with any infrastructure provider.

The PR includes also:
- a custom logger that stores the logs generated by clusterctl commands into log files.
- an InitManagementCluster method that implements a full bootstrap sequence: create a cluster, run clusterctl init and waits for providers to be ready

**Which issue(s) this PR fixes**:
xref #2753
xref #2637
xref #2636

/assign @vincepri
/assign @sedefsavas
/cc @wfernandes
/cc @gab-satchi

### CODE REVIEW

File: test/framework/clusterctl/client.go [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+		strings.Join(input.ControlPlaneProviders, ", "),
+		strings.Join(input.InfrastructureProviders, ", "),
+	))
+
+	initOpt := clusterctlclient.InitOptions{
+		Kubeconfig:              input.KubeconfigPath,
+		CoreProvider:            input.CoreProvider,
+		BootstrapProviders:      input.BootstrapProviders,
+		ControlPlaneProviders:   input.ControlPlaneProviders,
+		InfrastructureProviders: input.InfrastructureProviders,
```
Comment (sedefsavas) [inline test/framework/clusterctl/client.go:61]: nit. InfrastructureProviders can be a string instead of a list as in ConfigClusterInput.

File: test/framework/clusterctl/client.go [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+type MoveInput struct {
+	LogPath              string
+	ClusterctlConfigPath string
+	FromKubeconfigPath   string
+	ToKubeconfigPath     string
+	Namespace            string
+}
+
+// ClusterctlMove moves a workloads clusters.
+func ClusterctlMove(ctx context.Context, input MoveInput) {
```
Comment (sedefsavas) [inline test/framework/clusterctl/client.go:130]: nit. Rename to Move to be consistent with Init()?

File: test/framework/clusterctl/client.go [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+// MoveInput is the input for ClusterctlMove.
+type MoveInput struct {
+	LogPath              string
+	ClusterctlConfigPath string
+	FromKubeconfigPath   string
+	ToKubeconfigPath     string
+	Namespace            string
+}
+
+// ClusterctlMove moves a workloads clusters.
```
Comment (sedefsavas) [inline test/framework/clusterctl/client.go:129]:
```suggestion
// ClusterctlMove moves workload clusters.
```

File: test/framework/clusterctl/client.go [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+		Namespace:      input.Namespace,
+	}
+
+	Expect(clusterctlClient.Move(options)).To(Succeed(), "Failed to run clusterctl move")
+}
+
+func getClusterctlClientWithLogger(configPath, logName, logPath string) (clusterctlclient.Client, *logger.LogFile) {
+	log := logger.CreateLogFile(logger.CreateLogFileInput{
+		LogPath: logPath,
+		Name:    logName,
```
Comment (sedefsavas) [inline test/framework/clusterctl/client.go:148]: Do we want to keep clusterctl logs from different e2e runs like we do in docker e2e (it differentiates log file names by using container names)?
If so we may add some randomized addition at the end of logName.
This can be addressed in a follow-up enhancement PR.

File: test/framework/clusterctl/logger/logger.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+var _ logr.Logger = &logger{}
+
+func (l *logger) Enabled() bool {
+	return true
+}
+
+func (l *logger) Info(msg string, kvs ...interface{}) {
+	values := copySlice(l.values)
```
Comment (sedefsavas) [inline test/framework/clusterctl/logger/logger.go:44]: Can we use clusterctl/log package instead of adding a new one here?

File: test/framework/clusterctl/management_cluster.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+// Provides utilities for setting up a management cluster using clusterctl.
+
+// InitManagementClusterInput is the information required to initialize a new
+// management cluster for e2e testing.
+type InitManagementClusterInput struct {
+	// E2EConfig defining the configuration for the E2E test.
+	E2EConfig *E2EConfig
+
+	// ClusterctlConfigPath is the path to a clusterctl config file pointing to the repository to be used for running clusterctl init.
```
Comment (sedefsavas) [inline test/framework/clusterctl/management_cluster.go:42]:
```suggestion
	// ClusterctlConfigPath is the path to a clusterctl config file that points to repositories to be used during "clusterctl init".
```

File: test/framework/clusterctl/management_cluster.go [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+// InitManagementCluster returns a new cluster initialized and the path to the kubeConfig file to be used to access it.
+func InitManagementCluster(ctx context.Context, input *InitManagementClusterInput) (framework.ManagementCluster, string) {
+	// validate parameters and apply defaults
+
+	Expect(input.E2EConfig).ToNot(BeNil(), "Invalid argument. input.E2EConfig can't be nil when calling InitManagementCluster")
+	Expect(input.ClusterctlConfigPath).To(BeAnExistingFile(), "Failed to create a clusterctl config file at %s", input.ClusterctlConfigPath)
+
+	By(fmt.Sprintf("Creating the management cluster with name %s", input.E2EConfig.ManagementClusterName))
+
+	managementCluster, managementClusterKubeConfigPath, err := input.NewManagementClusterFn(input.E2EConfig.ManagementClusterName, input.Scheme)
```
Comment (sedefsavas) [inline test/framework/clusterctl/management_cluster.go:65]: Do we need to add a check here in case NewManagementClusterFn is nil?

Are we using kind.NewCluster as in the NewManagementClusterFn comment:

	// NewManagementClusterFn may be used to provide a custom function for
	// returning a new management cluster. Otherwise kind.NewCluster is used.

File: test/framework/clusterctl/client.go [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+	CoreProvider            string
+	BootstrapProviders      []string
+	ControlPlaneProviders   []string
+	InfrastructureProviders []string
+}
+
+// Init calls clusterctl init with the list of providers defined in the local repository
+func Init(ctx context.Context, input InitInput) {
+	By(fmt.Sprintf("clusterctl init --core %s --bootstrap %s --control-plane %s --infrastructure %s",
+		input.CoreProvider,
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply test/framework/clusterctl/client.go:50]: This question is related to the current kubeconfig file management, that is something that I would like to improve in a separate PR/after some considerations because it is potentially a breaking change.
TL;DR; Currently kubeconfig path is a random file under tmp folder, which makes it kind of unusable. Instead, IMO the kubeconfig path should be configurable, so we can store it in the artifact folder and used it for debugging.

File: test/framework/clusterctl/client.go (hunk 1 above)
Comment (fabriziopandini) [reply test/framework/clusterctl/client.go:61]: clusterctl fully support multitenancy, so the library must use lists because in case of multitenancy there can be more that one infrastructure provider in a Cluster.

File: test/framework/clusterctl/client.go (hunk 2 above)
Comment (fabriziopandini) [reply test/framework/clusterctl/client.go:130]: done!

File: test/framework/clusterctl/client.go (hunk 4 above)
Comment (fabriziopandini) [reply test/framework/clusterctl/client.go:148]: My assumption is that each test will produce more files than now (clusterctl logs, object dump & controller logs), and so files should be grouped in folders.
If this is not enough, we can consider making the file name configurable as well in a follow-up PR (random additions don't help you to understand e.g. the order of actions)

File: test/framework/clusterctl/logger/logger.go (hunk 5 above)
Comment (fabriziopandini) [reply test/framework/clusterctl/logger/logger.go:44]: Unfortunately not with the current implementation of the logger. 
TBH  I don't even consider this as a viable option, because when I proposed an extensible implementation for the clusterctl logger, it was turned down...
But eventually we can converge later if there is agreement on this

File: test/framework/clusterctl/management_cluster.go (hunk 7 above)
Comment (fabriziopandini) [reply test/framework/clusterctl/management_cluster.go:65]: fixed validation

**Comment by fabriziopandini** [comment]
@sedefsavas thanks for the feedback! comment addressed/answer provided

**Comment by sedefsavas** [comment]
Thanks! 
/lgtm
